### PR TITLE
L1 HSGS

### PR DESCRIPTION
**Pull request by liruipeng**
This PR provides a convergent l1-hybrid symmetric Gauss-Seidel (HSGS) method. It is named so since locally within the process it does l1-SGS, while across processes it does Jacobi-type updates, i.e., hybrid updates. The current l1-HSGS (of relaxation type 8) uses the same l1 term as in l1-hybrid Gauss-Seidel, which may not be convergent for general SPD matrices (e.g., the matrix included in PR). The new method, which uses the type number 88, has a new l1 weight that guarantees the convergence. Also, this PR provides the l1-symmetric hybrid Gauss-Seidel (SHGS), using the number 89, which is the symmetrized l1-HGS. The implementation is just simply calling the forward L1-HGS (13) followed by the backward one (14).
Based on hypre's regression tests, the convergence behavior of the new smoother is close to the old one, with up to 1 more iteration, for the problems that the old smoother can work.

- [x] tux
- [x] lassen
- [x] tioga
- [x] sunspot

### CODE REVIEW

File: src/parcsr_ls/par_amg_setup.c [hunk 1]
```diff
@@ -363,7 +363,7 @@ hypre_BoomerAMGSetup( void               *amg_vdata,
    /* probably should disable stuff like smooth num levels at some point */
 
 
-   if (grid_relax_type[0] >= 20) /* block relaxation choosen */
+   if (grid_relax_type[0] >= 20 && grid_relax_type[0] != 88) /* block relaxation choosen */
```
Comment (liruipeng) [inline src/parcsr_ls/par_amg_setup.c]: We run out of the numbers for relaxation that are less than to equal to 20. I picked 88 as the new full-fledged l1 SHGS. Maybe we should think about alternative ways for "rlx_type" other than number-based ways.

File: src/parcsr_ls/par_relax.c [hunk 2]
```diff
@@ -170,6 +170,12 @@ hypre_BoomerAMGRelax( hypre_ParCSRMatrix *A,
          hypre_BoomerAMGRelaxKaczmarz(A, f, omega, l1_norms, u);
          break;
 
+      case 88: /* L1 Symm. hybrid Gauss-Seidel */
```
Comment (liruipeng) [inline src/parcsr_ls/par_relax.c]: Note here that l1-SHGS (88) vs l1-HSGS (8).

File: src/parcsr_ls/ams.c [hunk 3]
```diff
@@ -700,6 +715,34 @@ HYPRE_Int hypre_ParCSRComputeL1Norms(hypre_ParCSRMatrix  *A,
...
+#else
+            HYPRE_THRUST_CALL( transform, l1_norm, l1_norm + num_rows, diag_tmp, l1_norm, l1_norm_op6() );
+#endif
+         }
+         else
+#endif
+         {
+            for (i = 0; i < num_rows; i++)
+            {
+               l1_norm[i] = (diag_tmp[i] + l1_norm[i] + hypre_sqrt(diag_tmp[i] * diag_tmp[i] + l1_norm[i] * l1_norm[i])) * 0.5;
```
Comment (liruipeng) [inline src/parcsr_ls/ams.c:741]: The essence of this PR is just this line.

**Comment by rfalgout** [comment]
Hi @liruipeng .  Thanks for putting this together!

It looks like you are switching relax type 8 to 88 and redefining type 8 (to have your new convergent weights).  I'm not sure we should change the behavior of type 8, but instead encourage our users to change to the new convergent method.  But, we should discuss as a team.

Also, type 8 was previously called in the user's manual "l1-scaled hybrid symmetric Gauss-Seidel".  Now in your comments it's called "L1 hybrid Symm. Gauss-Seidel" and type 88 is called "L1 Symm. hybrid Gauss-Seidel".

Can you edit the description at the top of this PR to summarize what you are doing, the names you are using for the smoothers, what each smoother actually does, and how it will change the behavior of hypre for users?  Basically, I would like to nail down the API documentation: https://github.com/hypre-space/hypre/blob/1d67edb3928e43c8ec0bcfaf55a2b7880b7a4b1b/src/parcsr_ls/HYPRE_parcsr_ls.h#L728-L752

Thanks!

**Comment by liruipeng** [comment]
> Hi @liruipeng . Thanks for putting this together!
> 
> It looks like you are switching relax type 8 to 88 and redefining type 8 (to have your new convergent weights). I'm not sure we should change the behavior of type 8, but instead encourage our users to change to the new convergent method. But, we should discuss as a team.
> 
> Also, type 8 was previously called in the user's manual "l1-scaled hybrid symmetric Gauss-Seidel". Now in your comments it's called "L1 hybrid Symm. Gauss-Seidel" and type 88 is called "L1 Symm. hybrid Gauss-Seidel".
> 
> Can you edit the description at the top of this PR to summarize what you are doing, the names you are using for the smoothers, what each smoother actually does, and how it will change the behavior of hypre for users? Basically, I would like to nail down the API documentation:
> 
> https://github.com/hypre-space/hypre/blob/1d67edb3928e43c8ec0bcfaf55a2b7880b7a4b1b/src/parcsr_ls/HYPRE_parcsr_ls.h#L728-L752
> 
> Thanks!

Thanks @rfalgout for the comment. I updated the description of the PR.

**Comment by rfalgout** [comment]
Hi @liruipeng .  I added this to the agenda for our meeting next week.  We definitely want to provide these smoothers for our users, but I want to be careful about how we do that.  Thanks for finding and fixing this problem!

Once we decide on a strategy, we'll need to update the header file API documentation.

**Comment by liruipeng** [comment]
> Hi @liruipeng . I added this to the agenda for our meeting next week. We definitely want to provide these smoothers for our users, but I want to be careful about how we do that. Thanks for finding and fixing this problem!
> 
> Once we decide on a strategy, we'll need to update the header file API documentation.

Sounds good to me. Thank you!

**Review by rfalgout — APPROVED** [review]
This looks good to me, Rui Peng.  We just need to update the API documentation in `HYPRE_parcsr_ls.h`.  Thanks!

File: src/parcsr_ls/HYPRE_parcsr_ls.h [hunk 4]
```diff
@@ -741,14 +741,32 @@ HYPRE_Int HYPRE_BoomerAMGSetGridRelaxType(HYPRE_Solver  solver,
...
+ *           for additional error approximation
  *    - 13 : \f$\ell_1\f$ Gauss-Seidel, forward solve
  *    - 14 : \f$\ell_1\f$ Gauss-Seidel, backward solve
  *    - 15 : CG (warning - not a fixed smoother - may require FGMRES)
  *    - 16 : Chebyshev
  *    - 17 : FCF-Jacobi
  *    - 18 : \f$\ell_1\f$-scaled jacobi
+ *    - 19 : Gaussian elimination (old version)
+ *    - 20 : Kaczmarz
+ *    - 21 : The same as 8 except forcing serialization on CPU (#OMP-thread = 1)
```
Comment (liruipeng) [inline src/parcsr_ls/HYPRE_parcsr_ls.h:761]: I wasn't able to find 21 in the code.

File: src/parcsr_ls/HYPRE_parcsr_ls.h [hunk 5]
```diff
@@ -741,14 +741,32 @@ HYPRE_Int HYPRE_BoomerAMGSetGridRelaxType(HYPRE_Solver  solver,
...
+ *    - 29 : Direct solve: use Gaussian elimination & BLAS
```
Comment (liruipeng) [inline src/parcsr_ls/HYPRE_parcsr_ls.h:762]: I wasn't able to find 29 in the code.

**Comment by liruipeng** [comment]
> This looks good to me, Rui Peng. We just need to update the API documentation in `HYPRE_parcsr_ls.h`. Thanks!

Thanks Rob. I updated the API documentation in the file.

File: src/parcsr_ls/HYPRE_parcsr_ls.h (hunk 5 above)
Comment (ulrikeyang) [reply src/parcsr_ls/HYPRE_parcsr_ls.h:762]: We have to be careful here:
Here are the definitions for the block relaxation schemes:
Switch statement to direct control based on relax_type:

    *     relax_type = 20 -> Jacobi or CF-Jacobi
    *     relax_type = 23 -> hybrid: SOR-J mix off-processor, SOR on-processor
    *                       with outer relaxation parameters (forward solve)
    *
    *     relax_type = 26 ->  hybrid: Jacobi off-processor,
    *                          Symm. Gauss-Seidel/ SSOR on-processor
    *                         with outer relaxation paramete
    *     relax_type = 29 -> Direct Solve

Comment (ulrikeyang) [reply src/parcsr_ls/HYPRE_parcsr_ls.h:762]: so Kaczmarz should probably not be 20 either.

File: src/parcsr_ls/par_amg_setup.c [hunk 6]
```diff
@@ -3288,11 +3291,27 @@ hypre_BoomerAMGSetup( void               *amg_vdata,
          }
       }
       else if (j == num_levels - 1 &&
-               (grid_relax_type[3] == 8 || grid_relax_type[3] == 13 || grid_relax_type[3] == 14))
+               (grid_relax_type[3] == 8 || grid_relax_type[3] == 89 || grid_relax_type[3] == 13 || grid_relax_type[3] == 14))
```
Comment (ulrikeyang) [inline src/parcsr_ls/par_amg_setup.c:3305]: Shouldn't 88 be included here as well, just as 8 is?

Comment (liruipeng) [reply src/parcsr_ls/par_amg_setup.c:3305]: Maybe not. 88 is handled below separately as it requires type 6 for `hypre_ParCSRComputeL1Norms`

File: src/parcsr_ls/HYPRE_parcsr_ls.h (hunk 5 above)
Comment (liruipeng) [reply src/parcsr_ls/HYPRE_parcsr_ls.h:762]: Thank you @ulrikeyang . I guess this issue (Kaczmarz type 20) has been existing for a while. Do you think it is just type=20 that has the issue or anything >= 20?

**Comment by liruipeng** [comment]
Hi @rfalgout @ulrikeyang I tried to resolve all the issues and pull in the latest `master` branch. Can you please take another look at this one? Thank you!

**Comment by liruipeng** [comment]
Hi @ulrikeyang. Can you please take another look at this one so we can merge it? Thank you!

**Review by ulrikeyang — APPROVED** [review]
Looks good. Thank you.

**Pull request merged**